design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+Then, for each asset, cdk-assets will perform the following operation:
+
+1. Assume the **publishing IAM role** in the target environment.
+2. Check if the asset is already published to this location. Assets are identified by their source hash. If it is, skip.
+3. If the asset doesn’t exists locally (e.g. docker image already exists, zip file already exists in local cache), package (docker build, zip directory).
+4. Publish the asset to the target location.
+
+In order for the publish to be able to execute `docker build`, this command must be executed in an environment that has docker available (in CodeBuild this means the project must be "privileged").
+
+**Templates as Assets**
+
+Some deployment systems (e.g. the CDK CLI) require that CloudFormation templates will be uploaded an S3 location before they can be deployed (this is done automatically in CodePipeline). Due to S3 eventual consistency, these files must be immutable, so we need to upload a new template file every time the template changes.
+
+To that end, we will treat all CloudFormation templates in the assembly like any other asset. They will be identified by their source hash (the hash of the template) and uploaded to the asset store in the environment in which they are expected to be deployed, like any other file asset.
+
+## Deployment
+
+At this point, all assets are published to asset stores in their target environments, so we can simply use standard CloudFormation deployment tools to deploy templates from the cloud assembly. Any references to assets were already resolved during synthesis.
+
+To deploy a stack to an environment, the deployment will need to:
+
+1. Assume the **Deployment IAM Role** from the target environment.
+2. Create a CloudFormation change-set for the stack.
+3. Execute the change-set by requesting CloudFormation to assume the administrative **CloudFormation IAM
+Role**.
+
+## Walkthrough
+
+In this section we will walk through the process of deploying a complex CDK app and how each one of the components in the toolchain is used within the workflow.
+
+### Bootstrapping
+
+We go to our ops team and ask them to prepare three AWS accounts for our application:
+
+1. `111111111DEP`: Deployment account
+2. `2222222222US`: US account
+3. `3333333333EU`: EU account
+
+The ops team will also bootstrap our environments:
+
+```shell
+$ cdk bootstrap aws://111111111DEP/us-west-2
+$ cdk bootstrap aws://2222222222US/us-east-1 --trust-account 111111111DEP
+$ cdk bootstrap aws://3333333333EU/eu-west-2 --trust-account 111111111DEP
+```
+
+> NOTE: each bootstrap command will be executed with the appropriate AWS credentials configured.
+
+The bootstrapping process will create the following resources:
+
+* 111111111DEP/us-west-2:
+  * `aws-cdk-files-111111111DEP-us-west-2`: S3 bucket
```

Comment:
None of these would be used. Is this a problem? Maybe we should not generate them if `--trust-account` has not been passed?